futures-core/src/async_trait/mod.rs

```diff
@@ -0,0 +1,210 @@
+    /// about the behavior of `poll` after an `Async` has completed.
+    ///
+    /// Callers who may call `poll` too many times may want to consider using
+    /// the `fuse` adaptor which defines the behavior of `poll`, but comes with
+    /// a little bit of extra cost.
+    fn poll(self: Pin<Self>, cx: &mut task::Context) -> Poll<Self::Output>;
+}
+
+impl<'a, F: ?Sized + Async> Async for &'a mut F {
+    type Output = F::Output;
+
+    fn poll(mut self: Pin<Self>, cx: &mut task::Context) -> Poll<Self::Output> {
+        unsafe { pinned_deref!(self).poll(cx) }
+    }
+}
+
+if_std! {
+    use std::boxed::{Box, PinBox};
+
+    impl<'a, F: ?Sized + Async> Async for Box<F> {
+        type Output = F::Output;
+
+        fn poll(mut self: Pin<Self>, cx: &mut task::Context) -> Poll<Self::Output> {
+            unsafe { pinned_deref!(self).poll(cx) }
+        }
+    }
+
+    impl<'a, F: ?Sized + Async> Async for PinBox<F> {
+        type Output = F::Output;
+
+        fn poll(mut self: Pin<Self>, cx: &mut task::Context) -> Poll<Self::Output> {
+            unsafe {
+                let this = PinBox::get_mut(Pin::get_mut(&mut self));
+                let re_pinned = Pin::new_unchecked(this);
+                re_pinned.poll(cx)
+            }
+        }
+    }
+
+    impl<'a, F: Async> Async for ::std::panic::AssertUnwindSafe<F> {
+        type Output = F::Output;
+
+        fn poll(mut self: Pin<Self>, cx: &mut task::Context) -> Poll<Self::Output> {
+            unsafe { pinned_field!(self, 0).poll(cx) }
+        }
+    }
+}
+
+/// An immediately-ready async value
+#[derive(Debug, Clone)]
+#[must_use = "async values do nothing unless polled"]
+pub struct Ready<T>(Option<T>);
+
+unsafe impl<T> Unpin for Ready<T> {}
+
+impl<T> Async for Ready<T> {
+    type Output = T;
+
+    fn poll(mut self: Pin<Self>, _cx: &mut task::Context) -> Poll<T> {
+        Poll::Ready(self.0.take().unwrap())
+    }
+}
+
+/// Create an immediately-ready async value.
+pub fn ready<T>(t: T) -> Ready<T> {
+    Ready(Some(t))
+}
+
+impl<F: Async> Async for Option<F> {
+    type Output = Option<F::Output>;
+
+    fn poll(mut self: Pin<Self>, cx: &mut task::Context) -> Poll<Self::Output> {
+        unsafe {
+            match Pin::get_mut(&mut self) {
+                None => Poll::Ready(None),
+                Some(ref mut x) => Pin::new_unchecked(x).poll(cx).map(Some),
+            }
+        }
+    }
+}
+
+/// A convenience for async `Result`s.
+pub trait AsyncResult {
+    /// The type of successful values
+    type Item;
+
+    /// The type of failures
+    type Error;
+
+    /// Poll this `AsyncResult` as if it were an `Async`.
+    ///
+    /// This method is a stopgap for a compiler limitation that prevents us from
+    /// directly inheriting from the `Async` trait; in the future it won't be
+    /// needed.
+    fn poll(self: Pin<Self>, cx: &mut task::Context) -> PollResult<Self::Item, Self::Error>;
```

Comment:
This should include `into_async` so that `AsyncResult`s can be `await`ed (well, I suppose it could go on `AsyncResultExt`, but the method should exist somewhere).